velox/expression/ComplexWriterTypes.h

```diff
@@ -1061,6 +1069,8 @@ class GenericWriter {
             ? "DynamicRow"
             : CppToType<ToType>::create()->toString());
 
+    casted_ = true;
```

Comment:
> Using the custom flags is to fix the bugs where `commit()` and `finish()` is called even if the GenericWriter is not casted.#6116
> 
> The reason behind the bugs is that `ensureWriter` is called not only during cast, but also sometimes in `init`. As a result, `castType_` will be set even if no any cast is done.
> 
> It will also be a problem here (and also in #6116) since I am directly write to the base vector without casting to any types. So `commit` and `finish` should not be called. Using custom flags solve the bugs


answered here

Reply:
Hi @Real-Chen-Happy,

> The reason behind the bugs is that ensureWriter is called not only during cast, but also sometimes in init. As a result, castType_ will be set even if no any cast is done.

I'm not sure if I understand the problem here. Could you elaborate a bit? I think when ensureWriter() is called, castType_ and castWriter_ are always set, so you can safely use them afterwards. Do you see this is not the case in some situations?

Reply:
The bug happened when I directly write into the vector held by GenericWriter. When this happened, no `commit` or `finish` should be called because I didn’t cast to any other writer type. If they are called then the vector internal state will be altered which will produce bugs. However, with current implementation castType_ will be set even if there is no cast to the other writer types. Consequently, `commit` or `finish` will be called. Let me know if my explain makes sense.

Reply:
castType_ is set sometimes during `init`, which causes the issues

Reply:
@Real-Chen-Happy If I understand correctly, basically when you write directly into the vector of GenericWriter, you don't want commit() and finish() to do anything, right? I don't think we can make commit() and finish() do nothing when ensureWriter() already set castType_ because this may break assumptions of other code using GenericWriter. 

I saw that without casting, castType_ is set in ensureWriter() only when vector type is primitive. Since you already call castTo\<V\> in your code when V is numeric, do you see the bug only when V is StringView?

Reply:
Yes. The bug only occur when the type is StringView

Reply:
> @Real-Chen-Happy If I understand correctly, basically when you write directly into the vector of GenericWriter, you don't want commit() and finish() to do anything, right? I don't think we can make commit() and finish() do nothing when ensureWriter() already set castType_ because this may break assumptions of other code using GenericWriter.

Makes sense. I think the new implementation where we serialize generic view inside `SingleValueAccumulator` and deserialize back to GenericView. And then using writer function `copyFrom` seems to avoid this issue. However, https://github.com/facebookincubator/velox/pull/6116 also seems to have a relevant issues. Do you think if we need any fix?

Reply:
Hi @Real-Chen-Happy, #6116 is a bit different. #6116 handles a situation where the GenericWriter is not casted and no data has been written to its vector, so it makes sense to skip commit().

For the problem you come across, one possible solution is to introduce StringWriter::setFinalized() to manually mark StringWriter::finalized_ to true so that when VectorWriter<Varchar>::commit() calls StringWriter::finalize(), it won't do anything.